File: matrix_spiral.py
#Time Complexity O(N^2)
#Space Complexity O(1)
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        rows = len(matrix)
        cols = len(matrix[0])
        result = []
        top , left = 0 , 0
        bottom = rows - 1
        right = cols - 1

        while (top <= bottom) and (left <= right):

            #Move from left to right of top row
            if (top <= bottom) and (left <= right):

                for i in range(left , right + 1):
                    result.append(matrix[top][i])
                top = top + 1

            #move from top to bottom of right col
            if (top <= bottom) and (left <= right):

                for i in range (top , bottom + 1):
                    result.append(matrix[i][right])
                right = right - 1


            # move from right to left of bottom row
            if (top <= bottom) and (left <= right):

                for i in range(right , left - 1 , -1):
                    result.append(matrix[bottom][i])
                bottom = bottom - 1

            # move from bottom to up of left col
            if (top <= bottom) and (left <= right):

                for i in range(bottom , top - 1, -1):
                    result.append(matrix[i][left])
                left = left + 1

        
        return result

File: Matrix_Spiral.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    /*
     * Time - O(m*n)
     * Space - O(1)
     */
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length, n = matrix[0].length;
        int top = 0, bottom = m-1, left = 0, right = n-1;
        List<Integer> result = new ArrayList<>();
        
        while (top<=bottom && left<=right) {
            // no need to check if top<=bottom since upper while condition will handle that
            for (int i=left; i<=right; i++) {
                result.add(matrix[top][i]);
            }
            top++;
            
            // no need to check if left<=right since after while condition above, we are not changing left and right
            for (int i=top; i<=bottom; i++) {
                result.add(matrix[i][right]);
            }
            right--;
            
            // check if top<=bottom
            if (top<=bottom) {
                for (int i=right; i>=left; i--) {
                    result.add(matrix[bottom][i]);
                }
            }
            bottom--;
            
            // check if left<=right
            if (left<=right) {
                for (int i=bottom; i>=top; i--) {
                    result.add(matrix[i][left]);
                }
            }
            left++;
        }
        return result;
    }
}